core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -44,81 +44,61 @@
-        return new WritePrimaryResult(request, result.getResponse(), result.getLocation(), primary);
-    }
-
-    @Override
-    protected final WriteReplicaResult shardOperationOnReplica(Request request, IndexShard replica) {
-        Translog.Location location = onReplicaShard(request, replica);
-        return new WriteReplicaResult(replica, request, location);
-    }
+    protected abstract WritePrimaryResult shardOperationOnPrimary(Request request, IndexShard primary) throws Exception;
 
     /**
-     * Simple result from a write action. Write actions have static method to return these so they can integrate with bulk.
+     * Called once per replica with a reference to the replica {@linkplain IndexShard} to modify.
+     *
+     * @return the result of the operation on replica, including current translog location and operation response and failure
+     * async refresh is performed on the <code>replica</code> shard according to the <code>ReplicaRequest</code> refresh policy
      */
-    public static class WriteResult<Response extends ReplicationResponse> {
-        private final Response response;
-        private final Translog.Location location;
-
-        public WriteResult(Response response, @Nullable Location location) {
-            this.response = response;
-            this.location = location;
-        }
-
-        public Response getResponse() {
-            return response;
-        }
-
-        public Translog.Location getLocation() {
-            return location;
-        }
-    }
+    @Override
+    protected abstract WriteReplicaResult shardOperationOnReplica(ReplicaRequest request, IndexShard replica) throws Exception;
 
     /**
      * Result of taking the action on the primary.
      */
-    class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
+    protected class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
         boolean finishedAsyncActions;
         ActionListener<Response> listener = null;
 
-        public WritePrimaryResult(Request request, Response finalResponse,
-                                  @Nullable Translog.Location location,
-                                  IndexShard indexShard) {
-            super(request, finalResponse);
-            /*
-             * We call this before replication because this might wait for a refresh and that can take a while. This way we wait for the
-             * refresh in parallel on the primary and on the replica.
-             */
-            new AsyncAfterWriteAction(indexShard, request, location, this, logger).run();
+        public WritePrimaryResult(ReplicaRequest request, @Nullable Response finalResponse,
+                                  @Nullable Location location, @Nullable Exception operationFailure,
+                                  IndexShard primary) {
+            super(request, finalResponse, operationFailure);
+            assert operationFailure != null ^ finalResponse != null;
```

Comment:
we can't asssert on location when operationFailure != null, as we can have noop update operation which will return a null location